Crafter Process Reminder and MIner Updated by redstone

AiDSl opened the pull request:
Crafter Process Reminder and MIner Updated by redstone

AiDSl committed:
Update IgneousMinerBlock.java

let it update

AiDSl committed:
Fix Reminder and Miner

Now Crafter also process reminders.
Fix Miner that did not update immediately when powered by redstone.

sorascode commented on src/main/java/com/herdlicka/igneousmachines/mixin/IgneousCrafterBlockEntityMixin.java:
```diff
@@ -1,33 +0,0 @@
-	@Unique
-	private static final ThreadLocal<ItemStack> REMAINDER_STACK = new ThreadLocal<>();
-
-	@Inject(method = "tick", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getItem()Lnet/minecraft/item/Item;"), locals = LocalCapture.CAPTURE_FAILHARD, allow = 1)
-	private static void getStackRemainder(World world, BlockPos pos, BlockState state, IgneousCrafterBlockEntity blockEntity, CallbackInfo ci) {
-		REMAINDER_STACK.set(blockEntity.getStack(9).getRecipeRemainder());
-	}
-
-	@ModifyArg(method = "tick", at = @At(value = "INVOKE", target = "Lnet/minecraft/util/collection/DefaultedList;set(ILjava/lang/Object;)Ljava/lang/Object;"), index = 1, allow = 1)
-	private static <E> E setStackRemainder(E element) {
-		E remainder = (E) REMAINDER_STACK.get();
-		REMAINDER_STACK.remove();
-		return remainder;
-	}
```
are you sure we want to remove this?

AiDSl replied:
yes, it essentially just a hack to make a blockentity variable that buffer the fuel item remainder, the reason of this is because in old logic, it immediately decrement/shrink the fuelStack and be come a empty stack, which will lose the remainder item infomation, now I buffed it in local var instead of doing it in member var with hacks.

sorascode commented on src/main/java/com/herdlicka/igneousmachines/block/entity/IgneousCrafterBlockEntity.java:
```diff
@@ -337,6 +344,34 @@ private static boolean craftRecipe(DynamicRegistryManager registryManager, @Null
+    {
+        for(int slot_index : slot_range)
+        {
+            ItemStack try_stack = slots.get(slot_index);
+            if( try_stack.isEmpty() )
+            {
+                slots.set(slot_index, item_to_insert.copy());
+                return;
+            }
+            if( ItemStack.canCombine(try_stack,item_to_insert) )
+            {
+                int combined_count=item_to_insert.getCount() + try_stack.getCount();
+                int stack_size=try_stack.getItem().getMaxCount();
+                if( combined_count <= stack_size )
+                {
+                    try_stack.setCount(combined_count);
+                    return;
+                }
+                try_stack.setCount(stack_size);
+                item_to_insert.setCount(combined_count-stack_size);
```
it would be nice if the behaviour was a little different. the vanilla item behaviour is that if there is an existing slot with that item then it adds it to that and only if there is none (or it overflows) it adds it to an empty slot

AiDSl replied:
k

sorascode commented on src/main/java/com/herdlicka/igneousmachines/block/entity/IgneousCrafterBlockEntity.java:
```diff
@@ -337,6 +344,34 @@ private static boolean craftRecipe(DynamicRegistryManager registryManager, @Null
+                item_to_insert.setCount(combined_count-stack_size);
+            }
+        }
+        ItemEntity item_to_drop=new ItemEntity(world,pos.getX(),pos.getY(),pos.getZ(),item_to_insert);
```
is this going it spawn it inside the crafter? thats seems a little unoptimal, would be nice if it was similar to the vanilla dropper where it spits it out

AiDSl replied:
K, will change.

sorascode commented on src/main/java/com/herdlicka/igneousmachines/block/entity/IgneousCrafterBlockEntity.java:
```diff
@@ -212,11 +213,12 @@ public static void tick(World world, BlockPos pos, BlockState state, IgneousCraf
                     if (hasFuel) {
-                        Item item = fuelStack.getItem();
+                        Item item_fuel = fuelStack.getItem();
+                        Item item_reminder=item_fuel.getRecipeRemainder();
```
typo: should be `item_remainder`

AiDSl replied:
ty

sorascode reviewed (changes requested):
dont forget to bump the version to a new minor

AiDSl committed:
Update IgneousCrafterBlockEntity.java

insertOrDrop now checking be stack size.
Flatten the nested tick logic, now it become 3 ifs : refuel, uncraft and crafting part

AiDSl committed:
I forgot to mention something :/

Depositer now react to redstone properly.
Crafter now drop item in its backside
--
I forgot to mention in last edit:
craftRecipe method ItemStack vars now named properly.
canAcceptRecipeOutput method now checked output tags too.

AiDSl committed:
edit changelog and version number :/

cherry on top

sorascode commented on gradle.properties:
```diff
@@ -9,7 +9,7 @@ yarn_mappings=1.20.1+build.2
 
 # Mod Properties
-mod_version=1.4.0
+mod_version=1.4.1
```
```suggestion
mod_version=1.5.0
```
should be minor version instead of patch

sorascode commented on CHANGELOG.md:
```diff
@@ -1,3 +1,5 @@
+* crafter code clean up.
+* crafter no longer eat your buckets.
+* crafter will drop item in its back side if inventory is full.
+* miner and depositer will update immediately.
```
```suggestion
* crafter will drop items from its back side if inventory is full.
* miner and depositer will update immediately when powered.
```

AiDSl committed:
Update gradle.properties

Co-authored-by: Sora <37141176+sorascode@users.noreply.github.com>

AiDSl committed:
Update CHANGELOG.md

Co-authored-by: Sora <37141176+sorascode@users.noreply.github.com>
